### PR TITLE
solhint runs in GitHub lint action and via package.json script

### LINKED ISSUE #27: Add solhint to CI
Pretty self-explanatory.

Adding solhint checks to the CI pipeline will help monitor and ensure basic quality of contracts.

### PR DESCRIPTION
**Pull request by aschlosberg**
Resolves #27

### CODE REVIEW

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -39,3 +39,12 @@ jobs:
...
         run: npx prettier '**/*.sol' --check
+
+      - name: Run solhint on production contracts
+        run: npx solhint 'contracts/**/*.sol'
+
+      - name: Run solhint on factory contracts
+        run: npx solhint 'factories/**/*.sol'
+
+      - name: Run solhint on test contracts
+        run: npx solhint 'tests/**/*.sol'
```
Comment (cxkoda) [inline .github/workflows/lint.yml:50]: Maybe merge this into a single call to the package script for consistency?
```
- name: Run solhint
  run: yarn solhint
```

**Review by cxkoda — CHANGES_REQUESTED** [review]
Thanks for adding this!

File: .github/workflows/lint.yml (hunk 1 above)
Comment (aschlosberg) [reply .github/workflows/lint.yml]: Good point!

File: .github/workflows/lint.yml [hunk 2]
```diff
@@ -40,11 +40,5 @@ jobs:
...
-      - name: Run solhint on production contracts
-        run: npx solhint 'contracts/**/*.sol'
-
-      - name: Run solhint on factory contracts
-        run: npx solhint 'factories/**/*.sol'
-
-      - name: Run solhint on test contracts
-        run: npx solhint 'tests/**/*.sol'
+      - name: Run solhint
+        run: test -z "$(yarn solhint)"
```
Comment (cxkoda) [inline .github/workflows/lint.yml:44]: Why do we need the `test -z` here? Shouldn't the `&&` concatenation in the script already yield the right exit code?

File: package.json [hunk 3]
```diff
@@ -6,6 +6,7 @@
   "scripts": {
     "build": "go build -o build/ethier ethier/*.go",
     "generate": "PATH=\"$(readlink -f build):${PATH}\" go generate ./...",
+    "solhint": "npx solhint 'contracts/**/*.sol' && npx solhint 'tests/**/*.sol' && npx solhint 'factories/**/*.sol'",
```
Comment (cxkoda) [inline package.json:9]: We could also replace this with `npx solhint '**/*.sol'` and add a `.solhintignore` file to exclude `node_modules`.
This way we don't have to worry about adding new folders if we add some in the future. Sorry for not thinking about this earlier.

**Review by cxkoda — CHANGES_REQUESTED** [review]
Thanks for the change.

I can't shake the feeling that `solhint` is doing nothing in our case. Afaik `solhint` should not be happy with inline assembly blocks, of which we have quite a few in our code. Will try this locally on my machine.

File: .github/workflows/lint.yml (hunk 2 above)
Comment (cxkoda) [reply .github/workflows/lint.yml:44]: I just noticed that this. `test -z` checks for empty strings, which we don't have with `solhint` afaik (prints a newline).

**Comment by cxkoda** [comment]
Now I understand why we never got any feedback from solhint - we are using a quite loose set of rules. I suggest we change `.solhint.json` to
```json
{
  "extends": "solhint:recommended",
  "rules": {
    "compiler-version": ["error", "^0.8.0"]
  }
}
```
This produces 101 warnings on my machine which we should address at some point.

I also suggest that we change the npm script to `solhint -w 0 '**/*.sol'` to make the CI also fail on warnings.

File: .github/workflows/lint.yml (hunk 2 above)
Comment (aschlosberg) [reply .github/workflows/lint.yml]: I was trying to figure out why we weren't getting any problems ¯\_(ツ)_/¯

**Comment by aschlosberg** [comment]
Thanks for basing that other PR on here! I'm very pro "warnings are errors" and am also in favour of a catch-all with ignore instead of the other way round (I've never used `solhint` before so didn't know this was a thing).

**Pull request merged**